### PR TITLE
Static2 themes

### PR DESCRIPTION
**Pull request by swergas**

### CODE REVIEW

File: assembl/static2/css/themes/default/_theme.scss [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+$title1-dark-color: $first-color;
+$title2-dark-color: $first-color;
+$title3-dark-color: $first-color;
+$title4-dark-color: $first-color;
+$title5-dark-color: $grey4;
+$title1-light-color: $white;
+$title2-light-color: $white;
+$title3-light-color: $white;
+$title4-light-color: $white;
+$title5-light-color: $white;
```
Comment (ayyazdaniaryan) [inline assembl/static2/css/themes/default/_theme.scss]: Will we be updating this file as we go along to include variables needed to make things overridable? Or are these the only things that we will be overriding?

File: assembl/static2/package.json [hunk 2]
```diff
@@ -9,8 +9,8 @@
   "scripts": {
     "test": "jest",
     "test:watch": "npm test -- --watch",
-    "dev": "webpack -d --watch",
-    "build": "webpack -p",
+    "dev": "webpack -d --watch --progress --colors",
```
Comment (ayyazdaniaryan) [inline assembl/static2/package.json]: I do not seem to have --progress in my version of webpack installed locally. Also, documentation doesn't seem to have --colors.

Comment (ayyazdaniaryan) [inline assembl/static2/package.json]: AHh whatever, it's just to prettify the output. no biggie...

File: assembl/templates/index_react.jinja2 [hunk 3]
```diff
@@ -2,7 +2,9 @@
 {% block page_title%}Assembl{% endblock%}
 
 {% block main_css %}
-	{# Override the main css. #}
+  {# Override the main CSS. #}
+  <link type="text/css" rel="stylesheet" href="/static2/build/bundle.css" />
```
Comment (ayyazdaniaryan) [inline assembl/templates/index_react.jinja2]: Do we have a bundle.css ?

Comment (ayyazdaniaryan) [inline assembl/templates/index_react.jinja2]: Wth is this?

File: assembl/static2/webpack.config.js [hunk 4]
```diff
@@ -1,11 +1,41 @@
...
+        paths = glob.sync('./css/themes/**/*_web.scss'),
+        i, path, parts, name;
+    for (i = 0; i < paths.length; i++) {
+        path = paths[i];
+        parts = path.split('/');
+        name = 'theme_' + parts[parts.length - 2] + '_web';
+        entries[name] = path;
+    }
+    paths = glob.sync('./css/themes/**/*_notifications.scss');
+    for (i = 0; i < paths.length; i++) {
```
Comment (ayyazdaniaryan) [inline assembl/static2/webpack.config.js]: node.js doesn't have a map or iter function baked in yet?

Comment (ayyazdaniaryan) [inline assembl/static2/webpack.config.js]: Ah you're using lodash. Makes this easier to read with a `forEach`. I'm nitpicky 😛

File: assembl/static2/webpack.config.js [hunk 5]
```diff
@@ -54,5 +84,6 @@ module.exports = {
           }
         }),
         new webpack.optimize.UglifyJsPlugin(),
+        new ExtractTextPlugin("[name].css"),
```
Comment (ayyazdaniaryan) [inline assembl/static2/webpack.config.js]: A) There isn't a documentation on pattern matching of name here. Is this a webpack thing?

B) Make a comment at top of this file that CSS changes are NOT hot-module friendly now.
https://github.com/webpack-contrib/extract-text-webpack-plugin/blob/webpack-1/README.md

File: assembl/static2/webpack.dev.config.js [hunk 6]
```diff
@@ -29,7 +30,8 @@ config.entry = {
     ]
 };
 config.plugins = [
-    new webpack.HotModuleReplacementPlugin()
+    new webpack.HotModuleReplacementPlugin(),
+    new ExtractTextPlugin("[name].css")
```
Comment (ayyazdaniaryan) [inline assembl/static2/webpack.dev.config.js]: Or does this fix the hot-module part??

File: assembl/templates/index_react.jinja2 (hunk 3 above)
Comment (ayyazdaniaryan) [inline assembl/templates/index_react.jinja2]: Okay, after reading the above, it's clear that this is from the extractor used above. Is this what that is??

File: assembl/static2/css/_assembl_base_styles.scss [hunk 7]
```diff
@@ -0,0 +1,27 @@
+//@import 'variables'; // The current theme imports the default theme which imports variables.scss. The current theme then overrides some of these variables.
```
Comment (ayyazdaniaryan) [inline assembl/static2/css/_assembl_base_styles.scss]: Add comment that any new css file *MUST* be added here to be compiled.

File: assembl/static2/package.json (hunk 2 above)
Comment (swergas) [inline assembl/static2/package.json]: I have --progress and --colors in my webpack version 1.14. The --progress flag is documented in http://webpack.github.io/docs/cli.html#progress . While you mention it, we should take care of the effect of these options in our build logs, and remove them if it creates information loss. But on my side for now I did not see any.

File: assembl/static2/css/themes/default/_theme.scss (hunk 1 above)
Comment (swergas) [inline assembl/static2/css/themes/default/_theme.scss]: Yes I should probably document more about the processes related to themes handling (how to install the new repo when upgrading a server, how to add new theme variables, etc). But the "how to create a new theme" process remains the same as in v1. Other processes are also almost identical.

File: assembl/templates/index_react.jinja2 (hunk 3 above)
Comment (swergas) [inline assembl/templates/index_react.jinja2]: Yes, "bundle" is the name of the main "entry" of webpack.config.js (see file). So webpack produces a bundle.js as well as a bundle.css, which contains CSS mentioned in all files which have been agregated in bundle.js.

File: assembl/static2/webpack.config.js (hunk 5 above)
Comment (swergas) [inline assembl/static2/webpack.config.js]: A) Yes, [name] is a pattern understood by webpack, see http://webpack.github.io/docs/configuration.html#output-filename

B) I'm not familiar at all with hot-module replacement, so I did not do any checks around this. I guess a theme system makes all hot reloading more difficult, that's almost unavoidable I guess.

File: assembl/static2/webpack.dev.config.js (hunk 6 above)
Comment (swergas) [inline assembl/static2/webpack.dev.config.js]: Honestly right now I don't remember the precise reason why I wrote this here

File: assembl/static2/webpack.config.js (hunk 5 above)
Comment (ayyazdaniaryan) [inline assembl/static2/webpack.config.js]: No, we never really thought that the hot-module replacement will be affected by the themefication. But at the very least, in the documentation, should add that and theme changes must be recompiled. It's a one-time processes...I hope.

File: assembl/static2/package.json (hunk 2 above)
Comment (ayyazdaniaryan) [inline assembl/static2/package.json]: No we're on the same version. Strange, I didn't see these options in the `--help`. But I did see it in the documentation online. I doubt that it affects the build logs at all. So all good 😄

File: assembl/static2/css/themes/README [hunk 8]
```diff
@@ -1,9 +1,30 @@
...
+* Use this variable in the CSS rules of your UI element. For example, in `static2/css/components/footer.scss`, add `.footer a { color: $footer-link-text-color; }`
+* Clear your browser cache, recompile CSS (for example using `npm run build` from within the `static2` folder), and see that result corresponds to your needs. Iterate if it's not the case yet.
+* If you think that this new variable is very likely to be overriden in futures themes, then copy the variable definition you wrote in `static2/css/variables.scss` to `static2/css/themes/default/_theme.scss`. As it is recommended that any new theme is copied from the default theme as starting point, any another new theme from now on will directly mention this variable and make it easy for the theme creator to override.
+* Overload the value of your new variable in another theme. For example, if you have already created the `sandbox` theme and if you store your themes in a repository called `your-themes-repository`, then copy the variable definition to `static2/css/themes/vendor/your-themes-repository/sandbox/_theme.scss`, and set a different color value for this variable.
+
```
Comment (ayyazdaniaryan) [inline assembl/static2/css/themes/README]: MUCH better. We can add to this as we continue our development...

**Pull request merged**